src/lib.rs

```diff
@@ -203,16 +208,215 @@ pub struct LockfileContent {
   // serializing this when it's empty
   /// Mapping between URLs and their checksums for "http:" and "https:" deps
   #[serde(default)]
-  remote: BTreeMap<String, String>,
+  pub(crate) remote: BTreeMap<String, String>,
   #[serde(skip_serializing_if = "WorkspaceConfigContent::is_empty")]
   #[serde(default)]
-  workspace: WorkspaceConfigContent,
+  pub(crate) workspace: WorkspaceConfigContent,
 }
 
 impl LockfileContent {
-  fn empty() -> Self {
+  pub fn from_json(
+    json: serde_json::Value,
+  ) -> Result<Self, DeserializationError> {
+    fn extract_nv_from_id(value: &str) -> Option<(&str, &str)> {
+      if value.is_empty() {
+        return None;
+      }
+      let at_index = value[1..].find('@').map(|i| i + 1)?;
+      let name = &value[..at_index];
+      let version = &value[at_index + 1..];
+      Some((name, version))
+    }
+
+    fn split_pkg_req(value: &str) -> Option<(&str, Option<&str>)> {
+      if value.len() < 5 {
+        return None;
+      }
+      // 5 is length of `jsr:@`/`npm:@`
+      let Some(at_index) = value[5..].find('@').map(|i| i + 5) else {
+        // no version requirement
+        // ex. `npm:jsonc-parser` or `jsr:@pkg/scope`
+        return Some((value, None));
+      };
+      let name = &value[..at_index];
+      let version = &value[at_index + 1..];
+      Some((name, Some(version)))
+    }
+
+    #[derive(Debug, Deserialize)]
+    struct RawNpmPackageInfo {
+      pub integrity: String,
+      #[serde(default)]
+      pub dependencies: Vec<String>,
+    }
+
+    #[derive(Debug, Deserialize)]
+    struct RawJsrPackageInfo {
+      pub integrity: String,
+      #[serde(default)]
+      pub dependencies: Vec<String>,
+    }
+
+    fn deserialize_section<T: DeserializeOwned + Default>(
+      json: &mut serde_json::Map<String, serde_json::Value>,
+      key: &'static str,
+    ) -> Result<T, DeserializationError> {
+      match json.remove(key) {
+        Some(value) => serde_json::from_value(value)
+          .map_err(|err| DeserializationError::FailedDeserializing(key, err)),
+        None => Ok(Default::default()),
+      }
+    }
+
+    use serde_json::Value;
+
+    let Value::Object(mut json) = json else {
+      return Ok(Self::empty_with_version("3".to_string()));
+    };
+
+    Ok(LockfileContent {
+      version: json
+        .remove("version")
+        .and_then(|v| match v {
+          Value::String(v) => Some(v),
+          _ => None,
+        })
+        .unwrap_or_else(|| "3".to_string()),
```

Comment:
Is this on purpose to default to version 3 instead of 4?

Reply:
Yes. This is only on with DENO_FUTURE for now